simulator/bypass/data_bypass.h

```diff
@@ -9,172 +9,251 @@
+inline auto operator""_RSG( unsigned long long int number)
+{
+    return RegisterStage( static_cast<uint8>( number));
+}
+
+
+
+template <typename ISA>
+class DataBypass
+{
+    using FuncInstr = typename ISA::FuncInstr;
+    using Register  = typename ISA::Register;
+
+    public:
         class BypassCommand
         {
             public:
-                BypassCommand( RegisterStage bypassing_stage, MIPSRegister register_num)
+                BypassCommand( RegisterStage bypassing_stage, Register register_num)
                     : bypassing_stage( bypassing_stage)
                     , register_num( register_num)
                 { }
 
                 auto get_bypassing_stage() const { return bypassing_stage; }
-                auto get_register_num() const { return register_num; }    
+                auto get_register_num() const { return register_num; }
 
             private:
                 const RegisterStage bypassing_stage;
-                const MIPSRegister register_num;
+                const Register register_num;
         };
 
+        // checks whether the source register of passed instruction is in RF  
+        auto is_in_RF( const FuncInstr& instr, uint8 src_index) const
+        {
+            const auto reg_num = instr.get_src_num( src_index);
+            return scoreboard.get_entry( reg_num).current_stage == RegisterStage::IN_RF;
+        }
+
+        // checks whether the source register of passed instruction is bypassible
+        auto is_bypassible( const FuncInstr& instr, uint8 src_index) const
+        {
+            const auto reg_num = instr.get_src_num( src_index);
+            return scoreboard.get_entry( reg_num).is_bypassible;
+        }
+
+        // checks if the stall needed for passed instruction
+        auto is_stall( const FuncInstr& instr) const
+        {
+            return ( (!is_in_RF( instr, 0) && !is_bypassible( instr, 0)) ||
+                     (!is_in_RF( instr, 1) && !is_bypassible( instr, 1)) );
+        }
+
+        // returns bypass command for passed instruction and its source register
+        // in accordance with current state of the scoreboard
+        auto get_bypass_command( const FuncInstr& instr, uint8 src_index) const
+        {
+            const auto reg_num = instr.get_src_num( src_index);
+            return BypassCommand( get_current_stage( reg_num), reg_num);
+        }
+
+        // returns an index of the port where bypassed data should be get from
+        // in accordance with passed bypass command
+        auto get_bypass_direction( const BypassCommand& bypass_command) const
+        {
+            const auto bypassing_stage = bypass_command.get_bypassing_stage();
+            return static_cast<uint8>( bypassing_stage);
+        }
+
+        // transforms bypassed data if needed in accordance with passed bypass command
+        static uint64 adapt_bypassed_data( const BypassCommand& bypass_command, uint64 bypassed_data);
```

Comment:
JFYI, it does not work for 64-bit DDIV instruction

Reply:
I think it should come with the constant 64-bit shift for all div/mult instructions.

Reply:
On top of that, `mips_multiplication` should have HI result invariably shifted on 64 bit regardless of the actual bitness of instruction.